File: speech_to_text/python/real_time/stream_radio.py
import json
import os
import threading

import requests
from websockets import ConnectionClosedOK
from websockets.sync.client import connect

# Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
api_key = os.environ.get("SONIOX_API_KEY")
websocket_url = "wss://stt-rt.soniox.com/transcribe-websocket"
audio_url = "https://npr-ice.streamguys1.com/live.mp3?ck=1742897559135"


def stream_audio(ws):
    with requests.get(audio_url, stream=True) as res:
        res.raise_for_status()
        for chunk in res.iter_content(chunk_size=4096):
            if chunk:
                ws.send(chunk)
    ws.send("")  # signal end of stream


def main():
    print("Opening WebSocket connection...")

    with connect(websocket_url) as ws:
        # Send start request
        ws.send(
            json.dumps(
                {
                    "api_key": api_key,
                    "audio_format": "auto",  # server detects the format
                    "model": "stt-rt-preview-v2",
                    "language_hints": ["en", "es"],
                }
            )
        )

        # Start streaming audio in background
        threading.Thread(target=stream_audio, args=(ws,), daemon=True).start()

        print(f"Transcription started from {audio_url}")

        final_text = ""

        try:
            while True:
                message = ws.recv()
                res = json.loads(message)

                if res.get("error_code"):
                    print(f"Error: {res['error_code']} - {res['error_message']}")
                    break

                non_final_text = ""

                for token in res.get("tokens", []):
                    if token.get("text"):
                        if token.get("is_final"):
                            final_text += token["text"]
                        else:
                            non_final_text += token["text"]

                print(
                    "\033[2J\033[H"  # clear the screen, move to top-left corner
                    + final_text  # write final text
                    + "\033[34m"  # change text color to blue
                    + non_final_text  # write non-final text
                    + "\033[39m"  # reset text color
                )

                if res.get("finished"):
                    print("\nTranscription complete.")
        except ConnectionClosedOK:
            pass
        except Exception as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    main()

File: speech_to_text/nodejs/real_time/stream_radio.js
import fetch from "node-fetch";

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
const apiKey = process.env.SONIOX_API_KEY;
const websocketUrl = "wss://stt-rt.soniox.com/transcribe-websocket";
const audioUrl = "https://npr-ice.streamguys1.com/live.mp3?ck=1742897559135";

// Connect to WebSocket API
const ws = new WebSocket(websocketUrl);

console.log("Opening WebSocket connection...");
ws.addEventListener("open", async () => {
  // Send start request
  ws.send(
    JSON.stringify({
      api_key: apiKey,
      audio_format: "auto", // server detects the format
      model: "stt-rt-preview-v2",
      language_hints: ["en", "es"],
    })
  );

  console.log("Transcription started.");

  try {
    // Connect to the radio stream
    const res = await fetch(audioUrl);
    if (!res.ok) {
      throw new Error(`HTTP error! Status: ${res.status}`);
    }
    res.body.on("data", (chunk) => {
      ws.send(chunk);
    });
    res.body.on("end", () => {
      console.log("Stream ended");
      ws.send(""); // signal end of stream
    });
    res.body.on("error", (error) => {
      console.error("Stream error:", error);
      ws.close();
    });
  } catch (error) {
    console.error("Error sending audio:", error);
    ws.close();
  }
});

let finalText = "";

ws.addEventListener("message", (event) => {
  // Receive and process messages
  const res = JSON.parse(event.data);

  if (res.error_code) {
    console.log(`\nError: ${res.error_code} ${res.error_message}`);
    process.exit(1);
  }

  let nonFinalText = "";

  for (const token of res.tokens || []) {
    if (token.text) {
      if (token.is_final) {
        finalText += token.text;
      } else {
        nonFinalText += token.text;
      }
    }
  }

  console.log(
    "\x1b[2J\x1b[H" + // clear the screen, move to top-left corner
      finalText + // write final text
      "\x1b[34m" + // change text color to blue
      nonFinalText + // write non-final text
      "\x1b[39m" // reset text color
  );

  if (res.finished) {
    console.log("\nTranscription done.");
  }
});

ws.addEventListener("close", () => {});

ws.addEventListener("error", (error) => {
  console.error("Connection error occurred:", error);
});
